reactivesocket-core/src/main/java/io/reactivesocket/internal/SafeCancellableSubscriber.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2016 Netflix, Inc.
+ * <p>
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ *  the License. You may obtain a copy of the License at
+ *  <p>
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *  <p>
+ *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations under the License.
+ */
+
+package io.reactivesocket.internal;
+
+import org.reactivestreams.Subscription;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+abstract class SafeCancellableSubscriber<T> extends CancellableSubscriber<T> {
+
+    protected final AtomicBoolean subscribed = new AtomicBoolean();
+    protected final AtomicBoolean done = new AtomicBoolean();
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        if (subscribed.compareAndSet(false, true)) {
+            super.onSubscribe(s);
+        } else {
+            onError(new IllegalStateException("Duplicate subscription."));
+        }
```

Comment:
Thanks @ktoso for the review!

Yes, one of my intentions behind this refactor was to centralize the creation of `Subscriber` and `Publisher` in the project so that we can easily verify the behaviors. Currently we create lambdas all over the code base which are hard to test and have bugs related to violating the spec (eg: #136 ). I intend to add tck validation tests for these subscribers and publishers to easily validate these conditions.